.github/scripts/format_dbt_test_failures.py

```diff
@@ -0,0 +1,256 @@
+# `dbt test --store-failures`, since it depends on two files created by
+# that operation (target/run_results.json and target/manifest.json).
+# It also requires Python dependencies be installed from
+# format_dbt_test_failures_requirements.txt.
+#
+# Accepts three optional positional arguments:
+#
+#   1. The local path to a run_results.json file generated by a test run
+#     * If not present, defaults to './target/run_results.json'
+#   2. The local path to a manifest.json file with the compiled dbt DAG
+#     * If not present, defaults to './target/manifest.json'
+#   3. The output filepath to which the workbook should be written
+#     * If not present, defaults to './qc_test_failures_<date>.xlsx'
+#
+# Expects one optional environment variable to be set:
+#
+#  1.AWS_ATHENA_S3_STAGING_DIR: Location in S3 where Athena query results
+#    should be written. If missing, defaults to the
+#    ccao-athena-results-us-east-1 bucket
+#
+# Each sheet in the output workbook represents a category of test, e.g.
+# "valid_range" or "not_null"; each row in a sheet represents a row in a
+# database that failed a test, with enough metadata that a reader can
+# figure out what conditions caused the test to fail and investigate the
+# root cause.
+#
+# Example usage:
+#
+#   AWS_ATHENA_S3_STAGING_DIR=s3://foo-bar-baz/ \
+#       python3 format_dbt_test_failures.py \
+#       ./target/run_results.json \
+#       ./target/manifest.json \
+#       ./qc_test_failures.xlsx
+
+import dataclasses
+import datetime
+import json
+import os
+import sys
+import typing
+
+import openpyxl
+import openpyxl.styles
+import pyathena
+import pyathena.cursor
+from openpyxl.worksheet.worksheet import Worksheet
+
+DEFAULT_TEST_CATEGORY = "miscellaneous"
+AWS_ATHENA_S3_STAGING_DIR = os.getenv(
+    "AWS_ATHENA_S3_STAGING_DIR", "s3://ccao-athena-results-us-east-1/"
+)
+
+
+@dataclasses.dataclass
+class FailedTestGroup:
+    """Class to store query results for a group of failed dbt tests and provide
+    convenience methods for formatting those results for output to a report."""
+
+    def __init__(self) -> None:
+        self._rows: typing.List[typing.Dict] = []
+
+    def update(self, new_rows: typing.List[typing.Dict]) -> None:
+        """Add a list of new_rows to the rows of failing tests tracked by
+        this group. The new_rows list should be formatted like the rows
+        returned by a csv.DictReader or a DictCursor, i.e. a list of
+        dicts mapping `{column_name: row_value}`."""
+        self._rows = [*self._rows, *new_rows]
+
+    @property
+    def fieldnames(self) -> typing.List[str]:
+        """Get a list of fieldnames that encapsulates all of the fieldnames
+        for all of the rows of failing tests tracked by this group."""
+        fieldnames = []
+        for row in self._rows:
+            for column in row.keys():
+                if column not in fieldnames:
+                    fieldnames.append(column)
+        return fieldnames
```

Comment:
**suggestion:** We should add a way to order this list with an arbitrary array of strings. ID columns should always be first in the workbook, so being able to pass `["taxyr", "parid", "card"]` and have those be the first 3 columns would be very useful. Caveat here is that not every test/group will have a `card` column.

Reply:
Good call, done in b1ee6eb1fb230a114d984063e79ce8f73fd7f8bc! It's possible we may want ways to change this order from sheet to sheet in the future, but for now the order is hardcoded and we can add an `__init__()` override later if we want.